src/network/core/tcp.cpp

```diff
@@ -138,6 +131,18 @@ SendPacketsState NetworkTCPSocketHandler::SendPackets(bool closing_down)
 	return SPS_ALL_SENT;
 }
 
+/**
+ * Get the function to transfer the data received from a socket into the packet.
+ * @param socket The socket to create the transfer function for.
+ * @return The function.
+ */
+static inline std::function<ssize_t(char*, size_t)> TransferTCPRecv(SOCKET &socket)
+{
+	return [&socket](char *data, size_t size) {
+		return recv(socket, data, static_cast<int>(size), 0);
+	};
+}
+
```

Comment:
Is using std::function really necessary here?
This introduces a lot of runtime indirection and so is a relative expensive abstraction.
It may be worth considering changing TransferIn/TransferOut to template functions so that this can be avoided.

Reply:
I don't know. I've never worked with std::function before and never heard of template functions. I guess I have some researching to do.

Reply:
I have made an attempt with templates. I'll let others be the judge whether templates are the way to go, or not. The first set of patches still adds the old std::function code, just the last to patches replace that with templates. So, depending on the consensus I either remove the last two patches, or I rebase the rest so the std::function code is not there anymore.

Reply:
I would consider templates better than using `std::function` (and `std::invoke`) wherever possible.

Note that you don't necessarily need to get rid of the lambdas, see the abstract-ish example at https://godbolt.org/z/qvd18T1jT (The `auto` return value is the trick there).

Reply:
IRC#openttd (~20-04-2021T19:05Z):
< frosch123> there is nothing wrong with your template solution
< frosch123> it's the same as the lambda version, just more raw code, instead of fancy code :)
< frosch123> ah, the transfer_function is only called once. yeah, stay with the templates
< frosch123> it's weird to capture stuff that is only used once

So I'll stay with the templated solution, unless someone really objects.